File: Sorting/CPP/insertion.cpp
#include <iostream>
#include <vector>

using namespace std;

class Insertion{
    public:
    static void sort(std::vector<int> &items){
        const int N = items.size();
        for (int i = 1; i < N; i++){
            for (int j = i; j > 0; j--){
                if (!Insertion::less(items[j], items[j-1])) break;
                Insertion::exch(items, j, j-1);
            }
            
        }
    }
    static bool less(int a, int b){
        return a < b;
    }
    static void exch(std::vector<int> &items, int i, int j){
        int temp = items[i];
        items[i] = items[j];
        items[j] = temp;
    }
};

int main(){
    std::vector<int> arr = {5, 4, 3, 2, 1};
    Insertion::sort(arr);
    for(auto item : arr){
		cout << item << " ";
    }
    return 0;
}

File: Sorting/Python/insertion.py
from typing import Any

class Insertion:
    @staticmethod
    def less(a: int, b: int) -> bool:
        return a < b
    
    @staticmethod
    def exch(items: Any, i: int, j: int) -> None:
        items[i], items[j] = items[j], items[i]
    
    @staticmethod
    def sort(items: Any) -> None:
        N = len(items)
        for i in range(N):
            for j in range(i, 0, -1):
                if not Insertion.less(items[j], items[j-1]):
                    break
                Insertion.exch(items, j, j-1)

if __name__ == '__main__':
    items = [5, 4, 3, 2, 1]
    Insertion.sort(items)
    print(items)

                
